libs/execution_context/ebpf_maps.c

```diff
@@ -215,9 +216,10 @@ _associate_program_with_prog_array_map(_In_ ebpf_core_map_t* map, _In_ const ebp
     // Validate that the program type is
     // not in conflict with the map's program type.
     const ebpf_program_type_t* program_type = ebpf_program_type(program);
-    if (program_array->program_type == NULL) {
-        program_array->program_type = program_type;
-    } else if (program_array->program_type != program_type) {
+    if (!program_array->is_program_type_set) {
+        program_array->is_program_type_set = TRUE;
+        program_array->program_type = *program_type;
```

Comment:
Do we need to serialize this update with a lock, in case multiple programs are loaded at the same time, which are using the same map?

Reply:
Updated

Reply:
Since there can be more than one type of programs in an ELF file, and associate will be called for all if them, this logic is going to pin the map with the first program that it gets associated with. But that may not be the program making the tail call at all. Why not set the program_type when the first program gets added to the prog_array map? It would be great if we could validate that the program calling the tail call has the same program type as the programs in the prog_array but  I am not sure how to do that ...

Reply:
An ELF file may have programs of different types that may refer to the same map. So associate will get called for all of them. Per the above implementation the map will end up with the program type of the first program it is associated with. But that program may not invoke tail_call and the other one might. IMO, the program type for a map should be associated with the program type of the first program whose FD was added by UM app. Now it will be additional goodness if we could further check that the program invoking the helper function has the same program type as associated with the prog_array.